Findemail

cliveleehere opened the pull request.

cliveleehere committed:
fix package name / directory

cliveleehere committed:
ignore build & out

cliveleehere committed:
- use WebCrawler factory so that additional parameters can be passed in
to the crawler constructors
- initial argument processors

cliveleehere committed:
Process argument for domain name

cliveleehere committed:
use default gradle java plugin directory

cliveleehere committed:
delete unused class

cliveleehere committed:
Email crawler

cliveleehere committed:
requirements.txt

mwomer-jana commented on src/main/java/com/clivelee/config/ArgumentProcessor.java:
```diff
@@ -0,0 +1,24 @@
+        URI uri = URI.create(urlString);
+
+        if (TextUtils.isEmpty(uri.getScheme())) {
+            urlString = "http://" + urlString;
```
This would seem to limit the search shouldVisit function to http only, or if the domain name (incorrectly) specified https, it would limit it to https.  Can you fix?

cliveleehere replied:
ah yes 👍

mwomer-jana commented on src/main/java/com/clivelee/crawler/Crawler.java:
```diff
@@ -0,0 +1,43 @@
+        controller.startNonBlocking(emailCrawlerFactory, numberOfCrawlers);
+
+        // Wait for 30 seconds
+        Thread.sleep(30 * 1000);
```
What's the justification here?  To keep from crawling indefinitely?

This would also seem to cause an invalid domain specified to take 30 seconds to exit.

cliveleehere replied:
ah, good point! will fix 👍

mwomer-jana commented on src/main/java/com/clivelee/crawler/EmailCrawler.java:
```diff
@@ -0,0 +1,49 @@
+    @Override
+    public boolean shouldVisit(Page referringPage, WebURL url) {
+        String href = url.getURL().toLowerCase();
+        return !FILTERS.matcher(href).matches()
```
This is overly restrictive and as such misses many pages linked from the start point of web.mit.edu.  Can you see the mistake and fix it?

cliveleehere replied:
hmm i suppose if the url has query parameters or fragments regex wouldn't match. or there may be other file types we might want as well, like htm.

mwomer-jana commented on src/test/java/com/clivelee/config/ArgumentProcessorTest.java:
```diff
@@ -0,0 +1,27 @@
+package com.clivelee.config;
+
+import org.junit.Test;
```
Are there more tests that weren't included?  How was the actual crawler tested?

cliveleehere replied:
i should've written more tests... i just ran the crawler to test it. :(

mwomer-jana commented on src/main/java/com/clivelee/crawler/EmailCrawler.java:
```diff
@@ -0,0 +1,49 @@
+            String text = htmlParseData.getText();
+            String html = htmlParseData.getHtml();
+
+            Set<String> foundEmails = TextFinder.findEmails(text + '\n' + html);
```
Why search both text and HTML?  What might be in one that's not in the other?

cliveleehere replied:
crawler4j was doing something funky under the hood. i think if i call `getHtml` for instances of HtmlPage and `getText` for other types of files, it should catch everything.

mwomer-jana commented on src/main/java/com/clivelee/crawler/EmailCrawler.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.regex.Pattern;
+
+final class EmailCrawler extends WebCrawler {
+    private final static Pattern FILTERS = Pattern.compile(".*(\\.(css|js|html))$");
```
Why CSS and JavaScript in the pattern?  The code for visit seems to restrict the email search to just HTML documents.

cliveleehere replied:
i'll include more file types

mwomer-jana commented on src/main/java/com/clivelee/utils/TextFinder.java:
```diff
@@ -0,0 +1,22 @@
+import static java.util.regex.Pattern.CASE_INSENSITIVE;
+
+public class TextFinder {
+    private static String emailRegex = "(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|\"(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21\\x23-\\x5b\\x5d-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])*\")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a- ...
```
Please explain this regex.

Just kidding.

cliveleehere replied:
😌

mwomer-jana reviewed (changes requested):
Hi Clive,
I left some comments, if you could address them please.

-M

cliveleehere committed:
added regex for url validation + tests
- apache common's url validator requires scheme, which we don't require

cliveleehere committed:
- Crawler refactor. use static factory method
- UrlFetchable chain-of-responsibility pattern to deal with URLs without
  scheme.
- try/catch for invalid configurations

cliveleehere committed:
pass domain String to EmailCrawlerFactory, so that the EmailCrawler can
look for an exact match of the domain

cliveleehere committed:
some more tests

cliveleehere commented:
Added a few more tests.

Been a while since i wrote a crawler and must say that was a lot of fun!

mwomer-jana commented on src/main/java/com/clivelee/crawler/EmailCrawler.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.regex.Pattern;
+
+final class EmailCrawler extends WebCrawler {
+    private final static Pattern FILTERS = Pattern.compile("([.](css|js|html|htm|pdf))([?].*)?$");
```
I was hoping for an explanation actually, but I was vague:  Why does the URL have to match this pattern?

It now seems to fix the case that I was looking for, which is URLs that end in / or otherwise don't have an extension. There's a lot of those on web.mit.edu that were being missed.  

I suspect your intention though was to factor in the optional query string portion of the URL, which was a good guess at what I was going for.  Can you explain what you were going for here, and/or take another crack at it?

mwomer-jana replied:
Ah, my original comment said "HTML documents", when I should have said something less ambiguous like: "HTML documents that end in .css, .html, .js, etc"

mwomer-jana commented on src/main/java/com/clivelee/Main.java:
```diff
@@ -0,0 +1,24 @@
+
+        try {
+            Crawler crawler = Crawler.create(configuration);
+            crawler.crawl();
```
When does this return?

mwomer-jana commented on src/main/java/com/clivelee/crawler/Crawler.java:
```diff
@@ -0,0 +1,72 @@
+        UrlFetchableChain startingChain = new UrlFetchableChain(pageFetcher, urlString);
+        UrlFetchableChain httpsChain = new UrlFetchableChain(pageFetcher, "https://" + urlString);
+        UrlFetchableChain httpChain = new UrlFetchableChain(pageFetcher, "http://" + urlString);
+
```
Ahh, I think these changes are a result of this comment: "This would seem to limit the search shouldVisit function to http only, or if the domain name (incorrectly) specified https, it would limit it to https. Can you fix?"

Which really was just to say that it was that by putting 'http://' on the front and using that string to compare with another URL to ensure you're not leaving the domain, would mean that you'd never be able to cross over into https from http.  That's fixed now via the changes to shouldVisit that pull the domain name out of the URL, rather than comparing it with scheme, et al.

mwomer-jana commented on src/main/java/com/clivelee/utils/TextFinder.java:
```diff
@@ -0,0 +1,20 @@
+
+public class TextFinder {
+    private static String emailRegex = "(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|\"(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21\\x23-\\x5b\\x5d-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])*\")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a- ...
+    private static Pattern p = Pattern.compile(emailRegex, Pattern.CASE_INSENSITIVE);
```
This seems to let in "addresses" like this: "//meche.mit.edu/people/faculty/LIENHARD@MIT.EDU" which it didn't in the previous version.

mwomer-jana reviewed (changes requested):
Thanks for the new code!  Tests are always good.

Can you explain what you're trying to do in shouldVisit?  I get that the return value determines whether it should continue to crawl or not.  I think you're determining that by the domain name being the same, and also that it ends in a known extension.  Is that right?

Also, this doesn't seem to exit!
